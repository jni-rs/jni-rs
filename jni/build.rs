extern crate walkdir;

use std::env;
use std::fs::symlink_metadata;
use std::path::{
    Path,
    PathBuf,
};

fn main() {
    if cfg!(feature = "invocation") {
        let libjvm_path = env::var("JAVA_HOME")
            .ok()
            .and_then(|p| find_libjvm(p))
            .or_else(|| find_java_home().and_then(|p| find_libjvm(p)));

        match libjvm_path {
            Some(path) => println!("cargo:rustc-link-search=native={}", path.display()),
            None => panic!("Failed to find libjvm.so. Try setting JAVA_HOME"),
        }

        println!("cargo:rustc-link-lib=dylib=jvm");
    }
}

fn find_libjvm<S: AsRef<Path>>(path: S) -> Option<PathBuf> {
    let walker = walkdir::WalkDir::new(path).follow_links(true);

    let expected_name = if cfg!(target_os = "windows") {
        "jvm.dll"
    } else if cfg!(target_os = "linux") {
        "libjvm.so"
    } else {
        "libjvm.dylib"
    };

    for entry in walker {
        let entry = match entry {
            Ok(entry) => entry,
            Err(_e) => continue,
        };

        let file_name = entry.file_name().to_str().unwrap_or("");

        if file_name == expected_name {
            return entry.path().parent().map(Into::into);
        }
    }

    None
}

fn find_java_home() -> Option<PathBuf> {
    let path = match env::var("PATH").ok() {
        Some(p) => p,
        None => return None,
    };

    let path_sep = if cfg!(target_os = "windows") {
        ";"
    } else {
        ":"
    };

    let paths = path.split(path_sep);
    let (mut exe_path, mut exe_meta): (PathBuf, _) = match paths
        .filter_map(|p| symlink_metadata(p).map(|m| (p.into(), m)).ok())
        .nth(0)
    {
        Some(v) => v,
        None => return None,
    };

    while exe_meta.file_type().is_symlink() {
        match (
            ::std::fs::read_link(&exe_path).ok(),
            symlink_metadata(&exe_path).ok(),
        ) {
            (Some(p), Some(m)) => {
                exe_path = p;
                exe_meta = m;
            }
            _ => return None,
        }
    }

    exe_path.parent().and_then(|p| p.parent()).map(Into::into)
}
