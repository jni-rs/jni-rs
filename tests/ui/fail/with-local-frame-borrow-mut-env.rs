#[path = "../../util/mod.rs"]
mod util;
use util::attach_current_thread;

use jni::Env;

pub fn main() {
    attach_current_thread(|env0: &mut Env| {
        env0.with_local_frame(10, |_env1: &mut Env| -> jni::errors::Result<_> {
            let _s = env0.new_string(c"hello").unwrap();
            eprintln!("BUG: this shouldn't compile since env0 shouldn't be mutable and new_string requires a mutable Env");
            Ok(())
        })
        .unwrap();
        Ok(())
    })
    .unwrap();
}
