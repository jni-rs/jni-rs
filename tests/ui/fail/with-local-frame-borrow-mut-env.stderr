error[E0500]: closure requires unique access to `*env0` but it is already borrowed
  --> tests/ui/fail/with-local-frame-borrow-mut-env.rs:9:35
   |
 9 |         env0.with_local_frame(10, |_env1: &mut Env| -> jni::errors::Result<_> {
   |         ---- ----------------     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ closure construction occurs here
   |         |    |
   |         |    first borrow later used by call
   |         borrow occurs here
10 |             let _s = env0.new_string(c"hello").unwrap();
   |                      ---- second borrow occurs due to use of `*env0` in closure
